src/presign/round_three.rs

```diff
@@ -63,44 +64,42 @@ pub(crate) struct Public {
 }
 
 impl Public {
+    /// Verify the validity of [`Public`] against the sender's [`AuxInfoPublic`]
+    /// and [`PublicBroadcast`](crate::presign::round_one::PublicBroadcast)
+    /// values.
+    ///
+    /// Note: The `receiver_...` values must be those of the _caller_ (i.e., the
+    /// verifier), and the `sender_...` values must be those of the other party
+    /// (i.e., the prover).
     pub(crate) fn verify(
         &self,
         context: &impl ProofContext,
-        receiver_keygen_public: &AuxInfoPublic,
-        sender_keygen_public: &AuxInfoPublic,
+        receiver_auxinfo_public: &AuxInfoPublic,
+        sender_auxinfo_public: &AuxInfoPublic,
         sender_r1_public_broadcast: &RoundOnePublicBroadcast,
     ) -> Result<()> {
         let mut transcript = Transcript::new(b"PiLogProof");
         let psi_double_prime_input = CommonInput::new(
             sender_r1_public_broadcast.K.clone(),
             self.Delta,
-            receiver_keygen_public.params().scheme().clone(),
-            sender_keygen_public.pk().clone(),
+            receiver_auxinfo_public.params().scheme().clone(),
+            sender_auxinfo_public.pk().clone(),
             self.Gamma,
         );
         self.psi_double_prime
             .verify(&psi_double_prime_input, context, &mut transcript)?;
 
         Ok(())
     }
+}
 
-    pub(crate) fn from_message(
-        message: &Message,
-        context: &impl ProofContext,
-        receiver_auxinfo_public: &AuxInfoPublic,
-        sender_auxinfo_public: &AuxInfoPublic,
-        sender_r1_public_broadcast: &RoundOnePublicBroadcast,
-    ) -> Result<Self> {
-        message.check_type(MessageType::Presign(PresignMessageType::RoundThree))?;
-        let round_three_public: Self = deserialize!(&message.unverified_bytes)?;
+impl TryFrom<&Message> for Public {
+    type Error = InternalError;
 
-        round_three_public.verify(
-            context,
-            receiver_auxinfo_public,
-            sender_auxinfo_public,
-            sender_r1_public_broadcast,
-        )?;
-        Ok(round_three_public)
+    fn try_from(message: &Message) -> std::result::Result<Self, Self::Error> {
+        message.check_type(MessageType::Presign(PresignMessageType::RoundThree))?;
+        let public: Self = deserialize!(&message.unverified_bytes)?;
```

Comment:
I'm thinking of adding the following to the documentation of `Public` for each round:
```
/// This type implements [`TryFrom`] on [`Message`], which validates that
/// [`Message`] is a valid serialization of `Public`, but _not_ that `Public` is
/// necessarily valid (i.e., that all the components are valid with respect to
/// each other); use [`Public::verify`] to check this latter condition.
```
Would this address your todo (maybe with some rewording of the comment)?

Reply:
Yeah, I like that doc.

My comment was based more on the fact that there are lots of types that have `from_message` and just deserialize and don't do any checks, but maybe should do checks. That's different from this case where we deserialize and don't do checks because there are no checks to do (and we audited that). But now that I think about it, that seems like something to handle via our issue tracking and not via comments in the code.